src/SpinWave/SpinWave.jl

```diff
@@ -0,0 +1,408 @@
+    end
+end
+
+"""
+    bogoliubov!
+
+Bogoliubov transformation that diagonalizes a bosonic Hamiltonian. 
+See Colpa JH. *Diagonalization of the quadratic boson hamiltonian* 
+Physica A: Statistical Mechanics and its Applications, 1978 Sep 1;93(3-4):327-53.
+"""
+function bogoliubov!(disp :: Vector{Float64}, V :: Matrix{ComplexF64}, Hmat :: Matrix{ComplexF64}, energy_tol :: Float64, mode_fast :: Bool = false)
+    @assert size(Hmat, 1) == size(Hmat, 2) "Hmat is not a square matrix"
+    @assert size(Hmat, 1) % 2 == 0 "dimension of Hmat is not even"
+
+    L = size(Hmat, 1) ÷ 2
+    (length(disp) != L) && (resize!(disp, L))
+
+    Σ = diagm([ones(ComplexF64, L); -ones(ComplexF64, L)])
+
+    if (!mode_fast)
+        eigval_check = eigen(Σ * Hmat).values
+        @assert all(<(energy_tol), abs.(imag(eigval_check))) "Matrix contains complex eigenvalues with imaginary part larger than `energy_tol`= "*string(energy_tol)*"(`sw_fields.coherent_states` not a classical ground state of the Hamiltonian)"
+
+        eigval_check = eigen(Hmat).values
+        @assert all(>(1e-12), real(eigval_check)) "Matrix not positive definite (`sw_fields.coherent_states` not a classical ground state of the Hamiltonian)"
+    end
+
+    K = cholesky(Hmat).U
+    @assert mode_fast || norm(K' * K - Hmat) < 1e-12 "Cholesky fails"
+
+    T = K * Σ * K'
+    eigval, U = eigen(Hermitian(T + T') / 2)
+
+    @assert mode_fast || norm(U * U' - I) < 1e-10 "Orthonormality fails"
+
+    # sort eigenvalues and eigenvectors
+    eigval = real(eigval)
+    # sort eigenvalues in descending order
+    index  = sortperm(eigval, rev=true)
+    eigval = eigval[index]
+    U = U[:, index]
+    for i = 1:2*L
+        if (i ≤ L && eigval[i] < 0.0) || (i > L && eigval[i] > 0.0)
+            error("Matrix not positive definite (`sw_fields.coherent_states` not a classical ground state of the Hamiltonian)")
+        end
+        pref = i ≤ L ? √(eigval[i]) : √(-eigval[i])
+        U[:, i] .*= pref
+    end
+
+    for col = 1:2*L
+        normalize!(U[:, col])
+    end
+
+    V[:] = K \ U
+
+    if (!mode_fast)
+        E_check = V' * Hmat * V
+        [E_check[i, i] -= eigval[i] for i = 1:L]
+        [E_check[i, i] += eigval[i] for i = L+1:2*L]
+        @assert all(<(1e-8), abs.(E_check)) "Eigenvectors check fails (Bogoliubov matrix `V` are not normalized!)"
+        @assert all(<(1e-6), abs.(V' * Σ * V - Σ)) "Para-renormalization check fails (Boson commutatition relations not preserved after the Bogoliubov transformation!)"
+    end
+
+    # The linear spin-wave dispersion also in descending order.
+    return [disp[i] = 2.0 * eigval[i] for i = 1:L]
+
+end
+
+"""
+    lswt_dispersion_relation
+
+Computes the spin excitation energy dispersion relations given a `SpinWaveField` and `k`. Note that `k` is a 3-vector, the units of kᵢ is 2π/|aᵢ|, where |aᵢ| is the lattice constant of the **chemical** lattice.
+"""
+function lswt_dispersion_relation(sw_fields :: SpinWaveFields, k :: Vector{Float64})
```

Comment:
I guess if it's a function that will be exposed to end-users, the brevity of the name `lswt_dispersion` has a certain benefit (vs a classical `dispersion` function?). A possible concern is that the name `spin_wave_dispersion` could imply either classical or linearized.

Reply:
I do like the brevity. I just wonder whether the full phrase "linear spin wave theory" is something that the user will automatically decode, or whether `lswt` it will just be a sequence of characters that has to be remembered. 

The "linearized" versus "not linearized" distinction could also break down somewhat if we one day ever add higher-order corrections to Hao's code. 

I understand that the classical dynamics can also be used to study spin waves in a sense, but I have the impression that "spin wave theory" does have an understood meaning in the magnetism community, referring to the quantized linearized dynamics and its extensions. What about just using `sw_dispersion` or `swt_dispersion`? 

I guess there's also a question of whether the existing function should be given a new prefix. We could have a "classical/cl" vs "spin_wave/sw" distinction throughout the code.

Just thinking out loud here.

Reply:
Good points. Your comment makes me wonder whether LSWT will become the default, and classical is the thing we prefix with `cl_`. I'm not sure at all, just trying to imagine how most people will be using this eventually. Given that loop corrections might also come, it does seem like the distinction is between classical and QM.

Reply:
After writing that out, I don't like the way `cl_` looks either. For `dispersion` there is no chance of ambiguity. For `intensities` the result will depend of the types given as input. Does this resolve it?

Reply:
Yes, I like that (i.e., just having `dispersion` and `intensities`, and having the latter dispatch on type.)